### PR TITLE
added implementation for fast serializable check for pandas write handler

### PR DESCRIPTION
**Pull request by dimosped**
Resolves issue #590

### CODE REVIEW

File: arctic/serialization/numpy_records.py [hunk 1]
```diff
@@ -12,11 +13,20 @@
...
         from pandas.tslib import Timestamp, get_timezone
 
+from ..exceptions import ArcticException
 
 log = logging.getLogger(__name__)
 
 DTN64_DTYPE = 'datetime64[ns]'
 
+# TODO: Switch on by default this flag to enable the fast check once this gets thoroughly tested
+_FAST_CHECK_DF_SERIALIZABLE = 'ENABLE_FAST_CHECK_DF_SERIALIZABLE' in os.environ
```
Comment (aflag) [inline arctic/serialization/numpy_records.py]: What if someone sets ENABLE_FAST_CHECK_DF_SERIALIZABLE=0?

File: arctic/serialization/numpy_records.py [hunk 2]
```diff
@@ -138,15 +148,48 @@ def _to_records(self, df, string_max_len=None, forced_dtype=None):
...
+            arr, _ = self._to_records(df.iloc[:1])  # only first row for performance
+            return arr, {}
+
+        # If only the Index has Objects,
+        # choose a small slice (two columns if possible, to avoid switching from a DataFrame to a Series)
+        df_objects_only = df[fields_with_object if fields_with_object else df.columns[:2]]
+
+        # Let any exceptions bubble up from here
+        arr, dtype = self._to_records(df_objects_only)
+        return arr, {f: dtype[f] for f in dtype.names}
```
Comment (aflag) [inline arctic/serialization/numpy_records.py:184]: I'm a bit concerned that the arr returned here could potentially pass one of the conditional checks in can_convert_to_records_without_objects when they really shouldn't.

File: arctic/serialization/numpy_records.py (hunk 1 above)
Comment (dimosped) [reply arctic/serialization/numpy_records.py]: yeah, this is a better convention instead: 
```python
"bool(os.environ.get('ENABLE_FAST_CHECK_DF_SERIALIZABLE'))"  
```

File: arctic/serialization/numpy_records.py (hunk 2 above)
Comment (dimosped) [reply arctic/serialization/numpy_records.py:184]: Is there a simple example of such a case which I could code in the form of a unit test ?
Which of the below is your main concern?

The conditional checks are two:
(1) If the resulting arr has objects (should be safe as we serialize whole columns of the df who contain objects)
```python
            if arr.dtype.hasobject:
                log.info('Pandas dataframe %s contains Objects, saving as Blob' % symbol)
                # Will fall-back to saving using Pickle
                return False
```

(2) Multi dimensional elements, we shouldn't do
```python
            elif any([len(x[0].shape) for x in arr.dtype.fields.values()]):
```

Comment (aflag) [reply arctic/serialization/numpy_records.py:184]: I was mainly concerned with (2). So you are saying that if len(x[0].shape) is larger than 0, then f_dtypes[f].hasobject will always be true? I think that makes sense. Would a np.array be considered an object in this case?

Comment (dimosped) [reply arctic/serialization/numpy_records.py:184]: I was under the assumption that this is the case, but is a good point.
I need to add a test to verify this, and prove compatibility.

Comment (dimosped) [reply arctic/serialization/numpy_records.py:184]: @aflag  I added test cases to cover this, see here for more details:
https://github.com/manahl/arctic/issues/590

**Review by aflag — APPROVED** [review]

**Comment by dimosped** [comment]
@bmoscon  this was the informal PR  for  https://github.com/manahl/arctic/issues/590 
to be merged on the main incremental serialization branch

**Comment by bmoscon** [comment]
looks good :)

**Pull request merged**